src/solver/expressions/include/antares/solver/expressions/Registry.hxx

```diff
@@ -0,0 +1,52 @@
+/*
+** Copyright 2007-2024, RTE (https://www.rte-france.com)
+** See AUTHORS.txt
+** SPDX-License-Identifier: MPL-2.0
+** This file is part of Antares-Simulator,
+** Adequacy and Performance assessment for interconnected energy networks.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the Mozilla Public Licence 2.0 as published by
+** the Mozilla Foundation, either version 2 of the License, or
+** (at your option) any later version.
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+#include <concepts>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+namespace Antares::Solver
+{
+//  Template class to manage the memory allocation and registry for a base class
+template<class Base>
+class Registry
+{
+public:
+    //  Method to create a new derived class object and add it to the registry
+    template<class Derived, class... Args>
+    requires std::derived_from<Derived, Base>
+    Base* create(Args&&... args)
```

Comment:
Registry::registry_ is a vector of unique_ptr<Base> and even if the returned ptr is a derived ptr, it would be necessarily used as Base* in the context of visitor::dispatch method

Reply:
Yes it's possible to return the derived type, I can make a PR if needed. It's slightly more convenient for the caller
### Current
```cpp
Node* add = registry.create<AddNode>(node1, node2);
Node* left = add->left(); // ERROR
Node* left = dynamic_cast<AddNode*>(add)->left(); // OK
```
### Proposed
```cpp
AddNode* add = registry.create<AddNode>(node1, node2); // Proposed
Node* left = add->left(); // OK
```

Reply:
Done, d91ed8a8d